lib/floki/html_tree.ex

```diff
@@ -328,4 +328,78 @@ defmodule Floki.HTMLTree do
       do_reduce(tree, fun.(head_node, acc), fun)
     end
   end
+
+  defimpl Inspect do
+    import Inspect.Algebra
+
+    def inspect(html_tree, opts) do
+      open = "%HTMLTree{"
+      close = "}"
+      container_opts = [separator: "", break: :flex]
+
+      container_doc(
+        open,
+        [{html_tree, html_tree.root_nodes_ids}],
```

Comment:
I think we could get the nodes and pass it right here, instead of traversing inside `fun/2`. It would carry the `html_tree` only for nodes that are of `HTMLNode` type.

Something like this:

```suggestion
        nodes_maybe_with_tree(html_tree, html_tree.root_nodes_ids),
```

And the function definition would be something like this:

```elixir
    defp nodes_maybe_with_tree(html_tree, nodes_ids) do
      nodes_ids
      |> Enum.reverse()
      |> Enum.map(fn node_id ->
        case Map.get(html_tree.nodes, node_id) do
          %HTMLNode{} = html_node ->
            {html_tree, html_node}

          other ->
            other
        end
      end)
    end
```
PS: It could be a `with` instead of a `case`.